packages/templates/fev4-fpm/src/page/index.ts

```diff
@@ -0,0 +1,102 @@
+import { join, dirname } from 'path';
+import { create as createStorage } from 'mem-fs';
+import { create, Editor } from 'mem-fs-editor';
+import { render } from 'ejs';
+
+import { CustomPage, CustomPageConfig } from '../types';
+
+/**
+ * Enhances the provided custom page configuration with default data
+ *
+ * @param {CustomPage} data - a custom page configuration object
+ */
+ export function enhanceData(data: CustomPage, manifestPath: string, fs: Editor): CustomPageConfig {
+    // enforce naming conventions
+    const firstChar = data.name[0];
+    const nameForId = firstChar.toLocaleLowerCase() + data.name.substring(1);
+    data.name = firstChar.toUpperCase() + data.name.substring(1)
+
+    const manifest: any = fs.readJSON(manifestPath);
+    const config: CustomPageConfig = {
+        ...data,
+        id: `${manifest['sap.app'].id}.ext.${nameForId}`,
+        path: join(dirname(manifestPath), 'ext', nameForId)
+    };
+    if (config.view === undefined) {
+        config.view = {
+            title: config.name
+        };
+    }
+    return config;
+}
+
+/**
+ * Validates the provided base path.
+ *
+ * @param {string} basePath - the base path
+ * @param {Editor} fs - the memfs editor instance
+ */
+export function validateBasePath(basePath: string, fs?: Editor) {
+    if (!fs) {
+        fs = create(createStorage());
+    }
+
+    const manifestPath = join(basePath, 'webapp', 'manifest.json');
+    if (!fs.exists(manifestPath)) {
+        throw new Error(`Invalid project folder. Cannot find required file ${manifestPath}`);
+    } else {
+        const manifest = fs.readJSON(manifestPath);
+        // TODO: check if this is an FE app
+    }
+
+}
+/**
+ * Writes the template to the memfs editor instance.
+ *
+ * @param {string} basePath - the base path
+ * @param {CustomPage} data - the custom page configuration
+ * @param {Editor} [fs] - the memfs editor instance
+ * @returns {Promise<Editor>} the updated memfs editor instance
+ */
+export async function generateCustomPage(basePath: string, data: CustomPage, fs?: Editor): Promise<Editor> {
+    if (!fs) {
+        fs = create(createStorage());
+    }
+    validateBasePath(basePath, fs);
+    
+    
+    const manifestPath = join(basePath, 'webapp/manifest.json');
+    const config = enhanceData(data, manifestPath, fs);
+    
+    // merge content into existing files
+    const root = join(__dirname, '../../templates/page');
+
+    // enhance manifest.json
+    fs.extendJSON(manifestPath, JSON.parse(render(fs.read(join(root, `manifest.json`)), config)), ( key, value ) => {
+        if (key === 'routes') {
+            const routes = value as object[];
+            routes.push({
+                pattern: `${config.navigation ? config.navigation.sourceEntity + '({key})/' + config.navigation.navEntity + '({key2})' : config.entity + '({key})'}:?query:`,
+                name: `${config.entity}${config.name}`,
+                target: `${config.entity}${config.name}`
```

Comment:
I am not sure how actual issue would be, but in case if FCL is enabled, then target property should be an array, like in following FCL example - https://sapui5.hana.ondemand.com/#/topic/7f651316d396400d88835adb0d021532

Reply:
Very good point, the current implementation would not work with FCL. I guess there is not way around checking the routing configuration, and if it is FCL, update all routes.

Reply:
I think technically the array notation also works for non FCL

Reply:
@815are could you check again, please?

Reply:
@tobiasqueck 
I think that logic is not correct - https://github.com/SAP/open-ux-tools/blob/fev4-fpm/packages/templates/fev4-fpm/src/page/index.ts#L52

I think it should be something like - 
```
if (sourceRoute?.target.constructor === Array) {
            newRoute.target = [...sourceRoute.target, (newRoute as Ui5Route).name];       
        } else {
            newRoute.target = newRoute.name;
        }
```

Proof is that in snapshot we see identical `target` for both routes - https://github.com/SAP/open-ux-tools/blob/fev4-fpm/packages/templates/fev4-fpm/test/unit/__snapshots__/page.test.ts.snap#L20-L31

```
"routes":  [
        {
          "name": "TestObjectPage",
          "pattern": "object/{key}",
          "target": [
            "TestObjectPage",
            "ChildEntityCustomPage",
          ],
        },
        {
          "name": "ChildEntityCustomPage",
          "pattern": "RootEntity({key})/navToChildEntity({key2}):?query:",
          "target": [
            "TestObjectPage",
            "ChildEntityCustomPage",
          ],
        },
      ]
```

As far as I know it should be:
```
"routes":  [
        {
          "name": "TestObjectPage",
          "pattern": "object/{key}",
          "target": [
            "TestObjectPage"
          ],
        },
        {
          "name": "ChildEntityCustomPage",
          "pattern": "RootEntity({key})/navToChildEntity({key2}):?query:",
          "target": [
            "TestObjectPage",
            "ChildEntityCustomPage",
          ],
        },
      ]
```

Reply:
Also in `spec` we have limitation for array length, something like:
```
// 1-3 level navigation is placed like [`1st Page`, `2nd Page`, `3rd Page`]
// 4+ level navigation is placed as fullscreen - [`4th+n Page`]
route.target = level <= 2 ? [...targetPages, route.target] : [route.target];
```
As I remember reason is that FCL supports only 3 columns - in result we do not need array with 4 columns.